x-pack/plugins/fleet/server/routes/security.ts

```diff
@@ -165,22 +93,199 @@ export async function getAuthzFromRequest(req: KibanaRequest): Promise<FleetAuth
+  }
+
+  return true;
+}
+
+interface FleetAuthzRouteConfig {
+  fleet?: Array<keyof FleetAuthz['fleet']>;
+  integrations?: Array<keyof FleetAuthz['integrations']>;
 }
 
-export const RouterWrappers: RouterWrappersSetup = {
-  require: {
-    superuser: (router) => {
-      return makeRouterEnforcingSuperuser(router);
+type FleetAuthzRouteRegistrar<
+  Method extends RouteMethod,
+  Context extends RequestHandlerContext = RequestHandlerContext
+> = <P, Q, B>(
+  route: FleetRouteConfig<P, Q, B, Method>,
+  handler: RequestHandler<P, Q, B, Context, Method>
+) => void;
+
+type FleetRouteConfig<P, Q, B, Method extends RouteMethod> = RouteConfig<P, Q, B, Method> & {
+  fleetAuthz?: FleetAuthzRouteConfig;
+  fleetRequireSuperuser?: boolean;
+  // TODO temporary required while agents call Fleet setup
+  fleetAllowFleetSetupPrivilege?: boolean;
+};
+
+// Fleet router that allow to add required access when registering route
+export interface FleetAuthzRouter<
+  TContext extends FleetRequestHandlerContext = FleetRequestHandlerContext
+> extends IRouter<TContext> {
+  get: FleetAuthzRouteRegistrar<'get', TContext>;
+  delete: FleetAuthzRouteRegistrar<'delete', TContext>;
+  post: FleetAuthzRouteRegistrar<'post', TContext>;
+  put: FleetAuthzRouteRegistrar<'put', TContext>;
+  patch: FleetAuthzRouteRegistrar<'patch', TContext>;
+}
+
+function shouldHandlePostAuthRequest(req: KibanaRequest) {
+  return req.route.path.match(/^\/api\/fleet/);
+}
+
+export function makeRouterWithFleetAuthz<TContext extends FleetRequestHandlerContext>(
+  router: IRouter<TContext>
+): { router: FleetAuthzRouter<TContext>; onPostAuthHandler: OnPostAuthHandler } {
+  const authzMap = new Map<
+    string,
+    {
+      fleetAuthz?: FleetAuthzRouteConfig;
+      fleetAllowFleetSetupPrivilege?: boolean;
+      fleetRequireSuperuser?: boolean;
+    }
+  >();
+
+  function routeKey(routeOptions: { path: string; method: string }) {
+    return `${routeOptions.method}:${routeOptions.path}`;
+  }
+
+  function addRouteAuthz(
+    routeOptions: { path: string; method: string },
+    authConfig: {
+      fleetAuthz?: FleetAuthzRouteConfig;
+      fleetAllowFleetSetupPrivilege?: boolean;
+      fleetRequireSuperuser?: boolean;
+    }
+  ) {
+    authzMap.set(routeKey(routeOptions), authConfig);
+  }
+
+  const fleetAuthzOnPostAuthHandler: OnPostAuthHandler = async (req, res, toolkit) => {
+    if (!shouldHandlePostAuthRequest(req)) {
+      return toolkit.next();
+    }
+
+    if (!checkSecurityEnabled()) {
+      return res.forbidden();
+    }
+
+    const fleetAuthzConfig = authzMap.get(routeKey(req.route));
+
+    if (!fleetAuthzConfig) {
+      return toolkit.next();
+    }
+    if (fleetAuthzConfig.fleetRequireSuperuser) {
+      if (!isSuperuser(req)) {
+        return res.forbidden({
+          body: {
+            message: SUPERUSER_AUTHZ_MESSAGE,
+          },
+        });
+      }
+      return toolkit.next();
+    }
+    const authz = await getAuthzFromRequest(req);
```

Comment:
We do not have access to the request handle context here, we could potentially use a WeakMap if we want to avoid a second call when building the request context

Reply:
The Elasticsearch has_privileges API that we call under the hood here is really fast, so I'm not very worried about this but it does mean more round-trips to ES in general. If the effort is fairly low (< 1hr), I think we should just do the WeakMap optimization now, otherwise I'm ok with deferring.

Reply:
I tried the `WeakMap` optimization and the req is not the same object between the handler and the `onPostAuth` hook, we could potentially use a cache with the req.uuid but I will defer it to a future optimization.